File: src/main/java/org/radarbase/appserver/service/questionnaire/protocol/SimpleReminderHandler.java
package org.radarbase.appserver.service.questionnaire.protocol;

import org.radarbase.appserver.dto.protocol.Assessment;
import org.radarbase.appserver.dto.protocol.ReminderTimePeriod;
import org.radarbase.appserver.dto.questionnaire.AssessmentSchedule;
import org.radarbase.appserver.dto.protocol.NotificationProtocol;
import org.radarbase.appserver.entity.Notification;
import org.radarbase.appserver.entity.Task;
import org.radarbase.appserver.entity.User;
import org.radarbase.appserver.service.questionnaire.notification.TaskNotificationGeneratorService;
import org.radarbase.appserver.service.questionnaire.notification.NotificationType;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.TimeZone;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class SimpleReminderHandler implements ProtocolHandler {
    private transient TaskNotificationGeneratorService taskNotificationGeneratorService = new TaskNotificationGeneratorService();
    private transient TimeCalculatorService timeCalculatorService = new TimeCalculatorService();

    public SimpleReminderHandler() { }

    @Override
    public AssessmentSchedule handle(AssessmentSchedule assessmentSchedule, Assessment assessment, User user) {
        TimeZone timezone = TimeZone.getTimeZone(user.getTimezone());
        NotificationProtocol protocol = assessment.getProtocol().getNotification();
        String title = this.taskNotificationGeneratorService.getTitleText(user.getLanguage(), protocol.getTitle(), NotificationType.REMINDER);
        String body = this.taskNotificationGeneratorService.getBodyText(user.getLanguage(), protocol.getBody(), NotificationType.REMINDER, assessment.getEstimatedCompletionTime());
        List<Notification> notifications = generateReminders(assessmentSchedule.getTasks(), assessment, timezone, user, title, body);
        assessmentSchedule.setReminders(notifications);
        return assessmentSchedule;
    }

    public List<Notification> generateReminders(List<Task> tasks, Assessment assessment, TimeZone timezone, User user, String title, String body) {
        return tasks.parallelStream()
                .flatMap(task -> {
                    ReminderTimePeriod reminders = assessment.getProtocol().getReminders();
                    return IntStream.range(0, reminders.getRepeat()).mapToObj(i -> {
                        Instant timestamp = timeCalculatorService.advanceRepeat(task.getTimestamp().toInstant(), reminders, timezone);
                        Notification notification = this.taskNotificationGeneratorService.createNotification(task, timestamp, title, body);
                        notification.setUser(user);
                        return notification;
                    });
                        }
                )
                .filter(notification-> (Instant.now().isBefore(notification.getScheduledTime()
                        .plus(notification.getTtlSeconds(), ChronoUnit.SECONDS)))).collect(Collectors.toList());
    }

}

File: src/main/java/org/radarbase/appserver/service/questionnaire/protocol/SimpleReminderHandler.kt
package org.radarbase.appserver.service.questionnaire.protocol

import org.radarbase.appserver.dto.protocol.Assessment
import org.radarbase.appserver.dto.questionnaire.AssessmentSchedule
import org.radarbase.appserver.entity.Notification
import org.radarbase.appserver.entity.Task
import org.radarbase.appserver.entity.User
import org.radarbase.appserver.service.questionnaire.notification.NotificationType
import org.radarbase.appserver.service.questionnaire.notification.TaskNotificationGeneratorService
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.*
import java.util.function.IntFunction
import java.util.stream.Collectors
import java.util.stream.IntStream

class SimpleReminderHandler : ProtocolHandler {
    @Transient
    private val taskNotificationGeneratorService = TaskNotificationGeneratorService()

    @Transient
    private val timeCalculatorService = TimeCalculatorService()

    override fun handle(
        assessmentSchedule: AssessmentSchedule,
        assessment: Assessment,
        user: User
    ): AssessmentSchedule {
        val timezone = TimeZone.getTimeZone(user.timezone)
        val protocol = assessment.protocol?.notification
        val assessmentScheduleTasks: List<Task>? = assessmentSchedule.tasks
        val estimatedCompletionTime = assessment.estimatedCompletionTime

        if (protocol == null || assessmentScheduleTasks == null || estimatedCompletionTime == null) {
            return assessmentSchedule.also {
                it.reminders = emptyList()
            }
        }

        val title =
            this.taskNotificationGeneratorService.getTitleText(user.language, protocol.title, NotificationType.REMINDER)
        val body = this.taskNotificationGeneratorService.getBodyText(
            user.language,
            protocol.body,
            NotificationType.REMINDER,
            estimatedCompletionTime
        )

        val notifications = generateReminders(
            assessmentScheduleTasks,
            assessment,
            timezone,
            user,
            title,
            body,
            protocol.email.enabled
        )
        assessmentSchedule.reminders = notifications
        return assessmentSchedule
    }

    fun generateReminders(
        tasks: List<Task>, assessment: Assessment, timezone: TimeZone,
        user: User, title: String, body: String, emailEnabled: Boolean
    ): List<Notification> {
        return tasks.parallelStream()
            .flatMap { task: Task ->
                val reminders = assessment.protocol?.reminders  ?: return@flatMap null
                val repeatReminders = reminders.repeat ?: return@flatMap null

                (0 until repeatReminders).map { _: Int ->
                    val timestamp = timeCalculatorService.advanceRepeat(task.timestamp!!.toInstant(), reminders, timezone)
                    taskNotificationGeneratorService.createNotification(
                        task, timestamp, title, body, emailEnabled
                    ).also {
                        it.user = user
                    }
                }.stream()
            }
            .filter { notification ->
                (Instant.now().isBefore(
                    notification.scheduledTime!!
                        .plus(notification.ttlSeconds.toLong(), ChronoUnit.SECONDS)
                ))
            }.collect(Collectors.toList())
    }
}
